utils/cache/filecache/filecache.go

```diff
@@ -0,0 +1,210 @@
+package filecache
+
+import (
+	"context"
+	"errors"
+	"sync/atomic"
+	"time"
+
+	"github.com/ARM-software/golang-utils/utils/commonerrors"
+	"github.com/ARM-software/golang-utils/utils/filesystem"
+	"github.com/ARM-software/golang-utils/utils/parallelisation"
+)
+
+type Cache struct {
+	entries       iEntryMap
+	entriesLM     iLockMap
+	entryProvider IEntryRetriever
+	fs            filesystem.FS
+	cfg           *FileCacheConfig
+	cancelStore   *parallelisation.CancelFunctionStore
+	closed        atomic.Bool
+}
+
+func (c *Cache) gc(ctx context.Context, _ time.Time) {
+	if err := parallelisation.DetermineContextError(ctx); err != nil {
+		return
+	}
+
+	c.entries.Range(func(key string, entry ICacheEntry) {
+		if entry.IsExpired() && c.entriesLM.TryLock(key) {
+			if err := entry.Delete(ctx); err == nil {
+				c.entries.Delete(key)
+				defer func() {
+					c.entriesLM.Unlock(key)
```

Comment:
what happens if an entry fails the delete, this won't get run. Then every tryLock will fail to lock.

Reply:
This is an example of where mocking would be good because you can have an explicit entry.Delete failure and see the behaviour and verify it won't happen